### PR TITLE
[SPARK-37719][BUILD] Remove the `-add-exports` compilation option introduced by SPARK-37070

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
In order to enable the UTs in `mllib-local` and `mllib` to use `mockito` to mock `j.u.Random`, `-add-exports=java.base/jdk.internal.util.random=ALL-UNNAMED` compile option is added for Java 17  in SPARK-37070 . 

This pr do the following change to remove the `-add-exports` option:
1. Introduce test dependence of mockito-inline, the new test dependence used to specify the configuration of `mockito-extensions`
2. Upgrade org.scalatestplus:mockito from 3-4 to 3-12 and upgrade mockito from 3.4.x to 3.12.x

### Why are the changes needed?
Find a way to let `mockito` can mock `j.u.Random` directly and remove `-add-exports` compile option for Java 17.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?

- Pass the Jenkins or GitHub Action

- Manual test with Java 17.0.1

Use maven
```
mvn clean install -pl mllib-local -am -DskipTests
mvn test -pl mllib-local 

mvn clean install -pl mllib -am -DskipTests
mvn test -pl mllib
```

Use sbt
```
build/sbt mllib-local/test
build/sbt mllib/test
```

There are no test failed related to ` org.mockito.exceptions.base.MockitoException: Mockito cannot mock this class: class java.util.Random.` without `-add-exports=java.base/jdk.internal.util.random=ALL-UNNAMED`, all test can be successful.

### CODE REVIEW

File: dev/deps/spark-deps-hadoop-2-hive-2.3 [hunk 1]
```diff
@@ -201,7 +201,7 @@ metrics-json/4.2.2//metrics-json-4.2.2.jar
 metrics-jvm/4.2.2//metrics-jvm-4.2.2.jar
 minlog/1.3.0//minlog-1.3.0.jar
 netty-all/4.1.68.Final//netty-all-4.1.68.Final.jar
-objenesis/2.6//objenesis-2.6.jar
+objenesis/3.2//objenesis-3.2.jar
```
Comment (LuciferYang) [inline dev/deps/spark-deps-hadoop-2-hive-2.3:204]: Let me find out why objenesis upgraded

Comment (LuciferYang) [reply dev/deps/spark-deps-hadoop-2-hive-2.3:204]: objenesis is the cascade dependency of mockito, it's scope is considered as `compile` in `unsafe` module:

```
[INFO] --- maven-dependency-plugin:3.1.1:tree (default-cli) @ spark-unsafe_2.12 ---
[INFO] org.apache.spark:spark-unsafe_2.12:jar:3.3.0-SNAPSHOT
[INFO] +- org.apache.spark:spark-tags_2.12:jar:3.3.0-SNAPSHOT:compile
[INFO] |  \- org.scala-lang:scala-library:jar:2.12.15:compile
[INFO] +- org.apache.spark:spark-tags_2.12:test-jar:tests:3.3.0-SNAPSHOT:test
[INFO] +- com.twitter:chill_2.12:jar:0.10.0:compile
[INFO] |  +- com.twitter:chill-java:jar:0.10.0:compile
[INFO] |  \- com.esotericsoftware:kryo-shaded:jar:4.0.2:compile
[INFO] |     \- com.esotericsoftware:minlog:jar:1.3.0:compile
[INFO] +- com.google.code.findbugs:jsr305:jar:3.0.0:compile
[INFO] +- com.google.guava:guava:jar:14.0.1:provided
[INFO] +- org.slf4j:slf4j-api:jar:1.7.30:provided
[INFO] +- org.mockito:mockito-core:jar:3.12.4:test
[INFO] |  +- net.bytebuddy:byte-buddy:jar:1.11.13:test
[INFO] |  +- net.bytebuddy:byte-buddy-agent:jar:1.11.13:test
[INFO] |  \- org.objenesis:objenesis:jar:3.2:compile
```

I haven't found the root cause yet.

Comment (dongjoon-hyun) [reply dev/deps/spark-deps-hadoop-2-hive-2.3:218]: I guess there are another libraries which use `objenesis` and this PR exposes a new version to supercede the old version of that library.

Comment (LuciferYang) [reply dev/deps/spark-deps-hadoop-2-hive-2.3:218]: Let me investigate again

Comment (LuciferYang) [reply dev/deps/spark-deps-hadoop-2-hive-2.3:218]: I will restart this work in a few days because I have something else to do

**Review by srowen — COMMENTED** [review]
Looks good to me if tests can pass

**Comment by LuciferYang** [comment]
merge with master and re-trigger Jenkins CI again

**Review by srowen — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
Although we don't know still why `objenesis` is upgraded here, I agree with @srowen 's decision.

**Comment by srowen** [comment]
Merged to master

**Comment by LuciferYang** [comment]
> Although we don't know still why objenesis is upgraded here, I agree with @srowen 's decision.

Sorry, I've been a little busy recently. I'll continue to investigate this when I have time @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
Never mind, @LuciferYang ~ It's okay to ignore that for now. ;)

**Pull request closed without merging**